.github/workflows/test.yml

```diff
@@ -0,0 +1,360 @@
+          tar -xzf /tmp/reports.tar.gz -C /tmp/reports
+          chown -R $(id -u):$(id -g) /tmp/reports
+          tree -nh /tmp/reports
+      -
+        name: Send to Codecov
+        uses: codecov/codecov-action@v3
+        with:
+          directory: ./bundles
+          env_vars: RUNNER_OS
+          flags: unit
+      -
+        name: Upload reports
+        if: always()
+        uses: actions/upload-artifact@v3
+        with:
+          name: unit-reports
+          path: /tmp/reports/*
+
+  docker-py:
+    runs-on: ubuntu-20.04
+    needs:
+      - build-dev
+    steps:
+      -
+        name: Checkout
+        uses: actions/checkout@v3
+      -
+        name: Set up runner
+        uses: ./.github/actions/setup-runner
+      -
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      -
+        name: Build dev image
+        uses: docker/bake-action@v2
+        with:
+          targets: dev
+          set: |
+            dev.cache-from=type=gha,scope=dev
+      -
+        name: Test
+        run: |
+          make -o build test-docker-py
+      -
+        name: Prepare reports
+        if: always()
+        run: |
+          mkdir -p bundles /tmp/reports
+          find bundles -path '*/root/*overlay2' -prune -o -type f \( -name '*-report.json' -o -name '*.log' -o -name '*.out' -o -name '*.prof' -o -name '*-report.xml' \) -print | xargs sudo tar -czf /tmp/reports.tar.gz
+          tar -xzf /tmp/reports.tar.gz -C /tmp/reports
+          chown -R $(id -u):$(id -g) /tmp/reports
+          tree -nh /tmp/reports
+      -
+        name: Test daemon logs
+        if: always()
+        run: |
+          cat bundles/test-docker-py/docker.log
+      -
+        name: Upload reports
+        if: always()
+        uses: actions/upload-artifact@v3
+        with:
+          name: docker-py-reports
+          path: /tmp/reports/*
+
+  integration-flaky:
+    runs-on: ubuntu-20.04
+    needs:
+      - build-dev
+    steps:
+      -
+        name: Checkout
+        uses: actions/checkout@v3
+      -
+        name: Set up runner
+        uses: ./.github/actions/setup-runner
+      -
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      -
+        name: Build dev image
+        uses: docker/bake-action@v2
+        with:
+          targets: dev
+          set: |
+            dev.cache-from=type=gha,scope=dev
+      -
+        name: Test
+        run: |
+          make -o build test-integration-flaky
+        env:
+          TEST_SKIP_INTEGRATION_CLI: 1
+
+  integration:
+    runs-on: ${{ matrix.os }}
+    needs:
+      - build-dev
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - ubuntu-20.04
+          - ubuntu-22.04
+        mode:
+          - ""
+          - rootless
+          - systemd
```

Comment:
We have to test cgroup v2 + systemd + rootless too, but can be revisited in another PR.

Reply:
Added in the matrix but some tests don't pass (ubuntu-22.04 + rootless + systemd): https://github.com/moby/moby/runs/8050470630?check_suite_focus=true#step:7:312

```
=== Failed
=== FAIL: amd64.integration.build TestCgroupNamespacesBuildDaemonHostMode (0.78s)
    build_cgroupns_linux_test.go:91: assertion failed: daemonCgroup is not containerCgroup

=== FAIL: amd64.integration.build TestBuildSquashParent (0.12s)
    build_squash_test.go:64: assertion failed: error is not nil: Error response from daemon: No such image: testbuildsquashparent:latest

=== FAIL: amd64.integration.build TestBuildMultiStageCopy/copy_to_root (0.12s)
    build_test.go:182: {"stream":"Step 1/6 : FROM busybox AS base"}

        {"stream":"\n"}

        {"stream":" ---\u003e 1c35c4412082\n"}

        {"stream":"Step 2/6 : RUN mkdir existingdir"}

        {"stream":"\n"}

        {"stream":" ---\u003e Running in 51df06f15e21\n"}

        {"stream":"Removing intermediate container 51df06f15e21\n"}

        {"errorDetail":{"message":"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory"},"error":"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory"}

        
    build_test.go:184: assertion failed: error is not nil: Error response from daemon: No such image: testbuildmultistagecopy/copy_to_root:latest
    --- FAIL: TestBuildMultiStageCopy/copy_to_root (0.12s)

=== FAIL: amd64.integration.build TestBuildMultiStageCopy/copy_to_newdir (0.12s)
    build_test.go:182: {"stream":"Step 1/8 : FROM busybox AS base"}

        {"stream":"\n"}

        {"stream":" ---\u003e 1c35c4412082\n"}

        {"stream":"Step 2/8 : RUN mkdir existingdir"}

        {"stream":"\n"}

        {"stream":" ---\u003e Running in 63743b53687f\n"}

        {"stream":"Removing intermediate container 63743b53687f\n"}

        {"errorDetail":{"message":"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory"},"error":"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory"}

        
    build_test.go:184: assertion failed: error is not nil: Error response from daemon: No such image: testbuildmultistagecopy/copy_to_newdir:latest
    --- FAIL: TestBuildMultiStageCopy/copy_to_newdir (0.12s)

=== FAIL: amd64.integration.build TestBuildMultiStageCopy/copy_to_newdir_nested (0.11s)
    build_test.go:182: {"stream":"Step 1/10 : FROM busybox AS base"}

        {"stream":"\n"}

        {"stream":" ---\u003e 1c35c4412082\n"}

        {"stream":"Step 2/10 : RUN mkdir existingdir"}

        {"stream":"\n"}

        {"stream":" ---\u003e Running in ede7a92110b6\n"}

        {"stream":"Removing intermediate container ede7a92110b6\n"}

        {"errorDetail":{"message":"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory"},"error":"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory"}

        
    build_test.go:184: assertion failed: error is not nil: Error response from daemon: No such image: testbuildmultistagecopy/copy_to_newdir_nested:latest
    --- FAIL: TestBuildMultiStageCopy/copy_to_newdir_nested (0.11s)

=== FAIL: amd64.integration.build TestBuildMultiStageCopy/copy_to_existingdir (0.12s)
    build_test.go:182: {"stream":"Step 1/12 : FROM busybox AS base"}

        {"stream":"\n"}

        {"stream":" ---\u003e 1c35c4412082\n"}

        {"stream":"Step 2/12 : RUN mkdir existingdir"}

        {"stream":"\n"}

        {"stream":" ---\u003e Running in 748a65cdf574\n"}

        {"stream":"Removing intermediate container 748a65cdf574\n"}

        {"errorDetail":{"message":"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory"},"error":"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory"}

        
    build_test.go:184: assertion failed: error is not nil: Error response from daemon: No such image: testbuildmultistagecopy/copy_to_existingdir:latest
    --- FAIL: TestBuildMultiStageCopy/copy_to_existingdir (0.12s)

=== FAIL: amd64.integration.build TestBuildMultiStageCopy/copy_to_newsubdir (0.12s)
    build_test.go:182: {"stream":"Step 1/14 : FROM busybox AS base"}

        {"stream":"\n"}

        {"stream":" ---\u003e 1c35c4412082\n"}

        {"stream":"Step 2/14 : RUN mkdir existingdir"}

        {"stream":"\n"}

        {"stream":" ---\u003e Running in d8527bb91ed5\n"}

        {"stream":"Removing intermediate container d8527bb91ed5\n"}

        {"errorDetail":{"message":"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory"},"error":"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory"}

        
    build_test.go:184: assertion failed: error is not nil: Error response from daemon: No such image: testbuildmultistagecopy/copy_to_newsubdir:latest
    --- FAIL: TestBuildMultiStageCopy/copy_to_newsubdir (0.12s)

=== FAIL: amd64.integration.build TestBuildMultiStageCopy (0.58s)

=== FAIL: amd64.integration.build TestBuildMultiStageOnBuild (2.09s)
    build_test.go:374: assertion failed: string "{\"stream\":\"Step 1/7 : FROM busybox AS stage1\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e 1c35c4412082\\n\"}\r\n{\"stream\":\"Step 2/7 : ONBUILD RUN echo 'foo' \\u003esomefile\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e Running in 5580a61188a1\\n\"}\r\n{\"stream\":\"Removing intermediate container 5580a61188a1\\n\"}\r\n{\"stream\":\" ---\\u003e 242dd34a6338\\n\"}\r\n{\"stream\":\"Step 3/7 : ONBUILD ENV bar=baz\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e Running in 76c6a1f6cc6a\\n\"}\r\n{\"stream\":\"Removing intermediate container 76c6a1f6cc6a\\n\"}\r\n{\"stream\":\" ---\\u003e e1de45dfc429\\n\"}\r\n{\"aux\":{\"ID\":\"sha256:e1de45dfc429d2f864649b73df5c9763ab9d3cd196b4a92781a9521f91d8025f\"}}\r\n{\"stream\":\"Step 4/7 : FROM stage1\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\"# Executing 2 build trigger\"}\r\n{\"stream\":\"s\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e Running in 2b2917e5c013\\n\"}\r\n{\"stream\":\"Removing intermediate container 2b2917e5c013\\n\"}\r\n{\"errorDetail\":{\"message\":\"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory\"},\"error\":\"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory\"}\r\n" does not contain "Successfully built"
    build_test.go:378: assertion failed: 3 (int) != 1 (int)

=== FAIL: amd64.integration.build TestBuildMultiStageLayerLeak (2.45s)
    build_test.go:483: assertion failed: string "{\"stream\":\"Step 1/8 : FROM busybox\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e 1c35c4412082\\n\"}\r\n{\"stream\":\"Step 2/8 : WORKDIR /foo\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e Running in 9d3037debf37\\n\"}\r\n{\"stream\":\"Removing intermediate container 9d3037debf37\\n\"}\r\n{\"stream\":\" ---\\u003e 827c099dabf5\\n\"}\r\n{\"stream\":\"Step 3/8 : COPY foo .\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e 3d5b41c4652f\\n\"}\r\n{\"aux\":{\"ID\":\"sha256:3d5b41c4652f91af1bbaf2d51ce75e90d81e901a2d92634ca643edb98ac07ba3\"}}\r\n{\"stream\":\"Step 4/8 : FROM busybox\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e 1c35c4412082\\n\"}\r\n{\"stream\":\"Step 5/8 : WORKDIR /foo\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e Using cache\\n\"}\r\n{\"stream\":\" ---\\u003e 827c099dabf5\\n\"}\r\n{\"stream\":\"Step 6/8 : COPY bar .\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e 4db0392ea7ed\\n\"}\r\n{\"stream\":\"Step 7/8 : RUN [ -f bar ]\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e Running in ed4477aada3d\\n\"}\r\n{\"stream\":\"Removing intermediate container ed4477aada3d\\n\"}\r\n{\"errorDetail\":{\"message\":\"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory\"},\"error\":\"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory\"}\r\n" does not contain "Successfully built"

=== FAIL: amd64.integration.build TestBuildWithHugeFile (0.13s)
    build_test.go:523: assertion failed: string "{\"stream\":\"Step 1/2 : FROM busybox\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e 1c35c4412082\\n\"}\r\n{\"stream\":\"Step 2/2 : RUN for g in $(seq 0 8); do dd if=/dev/urandom of=rnd bs=1K count=1 seek=$((1024*1024*g)) status=none; done \\u0026\\u0026     ls -la rnd \\u0026\\u0026 du -sk rnd\"}\r\n{\"stream\":\"\\n\"}\r\n{\"stream\":\" ---\\u003e Running in 7f7fd117c364\\n\"}\r\n{\"stream\":\"Removing intermediate container 7f7fd117c364\\n\"}\r\n{\"errorDetail\":{\"message\":\"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory\"},\"error\":\"open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory\"}\r\n" does not contain "Successfully built"
```

Will remove from the matrix for now and add a comment in the workflow.

Reply:
Looks like there's something to fix (look into) elsewhere as well. From this test, the error that occurs is;

```
open /sys/fs/cgroup/user.slice/user-1000.slice/cgroup.controllers: no such file or directory
```

And it looks like that error is converted somewhere to a `errdefs.ErrNotFound`
... which gets converted to a `404` / `No such image` in the API;

```
build_test.go:184: assertion failed: error is not nil: Error response from daemon: No such image: testbuildmultistagecopy/copy_to_newdir:latest
```

We probably need to change that into either a `errdefs.InvalidParam` (API: `400`), or an `errdefs.System` (API: `500`).

Reply:
Oh! No, I'm wrong; the "not found" is the check _after_ the build completed, which is triggered by the `apiclient.ImageInspectWithRaw()`. That's correct, because the image failed to build, so the image won't be there.

https://github.com/moby/moby/blob/402d106142e77a092dc761b354fb43d6935aa4f9/integration/build/build_test.go#L179-L184

The thing to improve in that test is to check the JSON output-stream, and to check if it has an `errorDetail` in the stream. The tricky bit with the `apiclient.ImageBuild` is that the `err` that's returned is only for the initial request, but it returns a JSON stream, and the actual _build_ failure will need to be checked from the stream that's returned (not the `err`). 

https://github.com/moby/moby/blob/402d106142e77a092dc761b354fb43d6935aa4f9/integration/build/build_test.go#L171-L177